File: 6_JobSequencing.c

#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int id, deadline, profit;
} Job;

int cmp(const void* a, const void* b) {
    return ((Job*)b)->profit - ((Job*)a)->profit;
}

void jobScheduling(Job jobs[], int n) {
    qsort(jobs, n, sizeof(Job), cmp);
    int result[n], slot[n];
    for (int i = 0; i < n; i++) slot[i] = 0;
    int count = 0, totalProfit = 0;
    for (int i = 0; i < n; i++) {
        for (int j = jobs[i].deadline - 1; j >= 0; j--) {
            if (!slot[j]) {
                result[j] = jobs[i].id;
                totalProfit += jobs[i].profit;
                slot[j] = 1;
                count++;
                break;
            }
        }
    }
    printf("Jobs done: %d, Total Profit: %d\n", count, totalProfit);
}

File: 6_JobSequencing.py
class Job:
    def __init__(self, id, deadline, profit):
        self.id = id
        self.deadline = deadline
        self.profit = profit

def jobScheduling(jobs, n):
    jobs.sort(key=lambda x: x.profit, reverse=True)
    max_deadline = max(job.deadline for job in jobs)
    slots = [-1] * (max_deadline + 1)
    count, total_profit = 0, 0
    for job in jobs:
        for j in range(job.deadline, 0, -1):
            if slots[j] == -1:
                slots[j] = job.id
                count += 1
                total_profit += job.profit
                break
    return count, total_profit
